### PR TITLE
internal: Object Lazy Stat

### PR DESCRIPTION
**Pull request by JinnyYi**
fix: https://github.com/beyondstorage/site/issues/115

### CODE REVIEW

File: docs/go-storage/internal/how-types.NewObject-works.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+---
+title: How types.NewObject(client, done) works
```
Comment (Xuanwo) [inline docs/go-storage/internal/how-types.NewObject-works.md:2]: How about using the title `Object Lazy Stat`?

File: docs/go-storage/internal/how-types.NewObject-works.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
+---
+title: How types.NewObject(client, done) works
+---
+
+### Introduction
+
+We use `types.NewObject(client, done)` to create a new object with client in services. 
+
+> func NewObject(client Storager, done bool) *Object
```
Comment (xxchan) [inline docs/go-storage/internal/how-types.NewObject-works.md:9]: Use code block like
````
```go
func NewObject(client Storager, done bool) *Object
```
````

File: docs/go-storage/internal/how-types.NewObject-works.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+We use `types.NewObject(client, done)` to create a new object with client in services. 
+
+> func NewObject(client Storager, done bool) *Object
+
+**Parameters**
+
+client - the client in which `Object` is alive.
+
+done - marks whether the object stated. `true` specifies this object already stated, we don't need to `stat` it anymore.
```
Comment (xxchan) [inline docs/go-storage/internal/how-types.NewObject-works.md:15]: What about `stat`ed ?

File: docs/go-storage/internal/how-types.NewObject-works.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+**Return**
+
+An `Object` pointer points to the object carries all object metadata.
+
+`Object` carries all the object metadata, including the client in witch the object alive, global metadata for object, service defined metadata, user defined metadata. And it survives `get/set` functions. Also, all the metadata could be got by `stat` internal.
```
Comment (Prnyself) [inline docs/go-storage/internal/how-types.NewObject-works.md:21]: `survives` or `support`?

File: docs/go-storage/internal/how-types.NewObject-works.md (hunk 3 above)
Comment (JinnyYi) [reply docs/go-storage/internal/how-types.NewObject-works.md]: `stat` for the object.
I've updated the description, please help to check.

File: docs/go-storage/internal/object-lazy-stat [hunk 5]
```diff
@@ -0,0 +1,44 @@
+---
+title: Object Lazy Stat
+---
+
+### Introduction
```
Comment (Xuanwo) [inline docs/go-storage/internal/object-lazy-stat:5]: Use `## Introduction` instead.

File: docs/go-storage/internal/object-lazy-stat [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+done - marks whether `stat` for the object has been performed. `true` indicates `stat` has been performed, we don't need to `stat` for this object anymore.
+
+**Return**
+
+An `Object` pointer points to the object carries all object metadata.
+
+`Object` carries all the object metadata, including the client in witch the object alive, global metadata for object, service defined metadata, user defined metadata. And it supports `get/set` functions. Also, all the metadata could be got by `stat` internal.
+
+### Lazy Stat Strategy
```
Comment (Xuanwo) [inline docs/go-storage/internal/object-lazy-stat:25]: ditto.

File: docs/go-storage/internal/object-lazy-stat [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+`Lazy Stat` strategy is to avoid `stat` too many times. It's used to fetch data from the remote when required and no matter `stat` success or not, it will be executed exactly once.
+
+Refer to sync.Once, variable `done` is introduced to mark the execution state of `stat`. `sync.Mutex` and `sync.atomic` are used to ensure read `done` in a thread-safe way.
+
+To get a certain object metadata from an `Object` instance, `stat` belongs to `Object` will be called first. `stat` calls the function `Stat` belongs to the `client` in `Object` if the `stat` is being called for the first time for this instance of `Object`. In other words, given var `object Object`, if `object.stat()` is called multiple times, only the first call will invoke `object.client.Stat()`.
+
+### Instructions
```
Comment (Xuanwo) [inline docs/go-storage/internal/object-lazy-stat:33]: ditto.

**Comment by xxchan** [comment]
Why the page is missing in the deployment? Where is it?

**Comment by Xuanwo** [comment]
> Why the page is missing in the deployment? Where is it?

I think the file should be ended with `.md`.

File: docs/go-storage/internal/object-lazy-stat [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+client - the client in which `Object` is alive.
+
+done - marks whether `stat` for the object has been performed. `true` indicates `stat` has been performed, we don't need to `stat` for this object anymore.
+
+**Return**
+
+An `Object` pointer points to the object carries all object metadata.
+
+`Object` carries all the object metadata, including the client in witch the object alive, global metadata for object, service defined metadata, user defined metadata. And it supports `get/set` functions. Also, all the metadata could be got by `stat` internal.
```
Comment (xxchan) [inline docs/go-storage/internal/object-lazy-stat:23]: Typo : witch

File: docs/go-storage/internal/object-lazy-stat.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+```go
+func NewObject(client Storager, done bool) *Object
+```
+
+**Parameters**
+
+client - the client in which `Object` is alive.
+
+done - marks whether `stat` for the object has been performed. `true` indicates `stat` has been performed, we don't need to `stat` for this object anymore.
```
Comment (xxchan) [inline docs/go-storage/internal/object-lazy-stat.md:17]: When will we create an object on which `stat` has been performed? I'm a little bit confused

**Review by Xuanwo — APPROVED** [review]

**Pull request merged**